src/faq/troubleshooting.rst

```diff
@@ -0,0 +1,222 @@
+  document version
+
+In this instance, there are two possible failure scenarios to think about. If
+the most recent document version has a quantity of 0, you handle it just like
+you would in a RDBMS and alert the user that they can't actually buy what they
+wanted to purchase. If the most recent document version has a quantity greater
+than 0, you simply repeat the operation with the updated data, and start back
+at the beginning. This forces you to do a bit more work than an RDBMS would, and
+could get a little annoying if there are frequent, conflicting updates.
+
+Now, the answer I just gave presupposes that you're going to do things in
+CouchDB in much the same way that you would in an RDBMS. I might approach this
+problem a bit differently:
+
+I'd start with a "master product" document that includes all the descriptor data
+(name, picture, description, price, etc). Then I'd add an "inventory ticket"
+document for each specific instance, with fields for product_key and claimed_by.
+If you're selling a model of hammer, and have 20 of them to sell, you might have
+documents with keys like hammer-1, hammer-2, etc, to represent each available
+hammer.
+
+Then, I'd create a view that gives me a list of available hammers, with a reduce
+function that lets me see a "total". These are completely off the cuff, but
+should give you an idea of what a working view would look like.
+
+**Map**::
+
+    function(doc)
+    {
+        if (doc.type == 'inventory_ticket' && doc.claimed_by == null ) {
+            emit(doc.product_key, { 'inventory_ticket' :doc.id, '_rev' : doc._rev });
+        }
+    }
+
+This gives me a list of available "tickets", by product key. I could grab a
+group of these when someone wants to buy a hammer, then iterate through sending
+updates (using the id and _rev) until I successfully claim one (previously
+claimed tickets will result in an update error).
+
+**Reduce**::
+
+    function (keys, values, combine) {
+        return values.length;
+    }
+
+This reduce function simply returns the total number of unclaimed
+inventory_ticket items, so you can tell how many "hammers" are available for
+purchase.
+
+**Caveats**
+
+This solution represents roughly 3.5 minutes of total thinking for the
+particular problem you've presented. There may be better ways of doing this!
+That said, it does substantially reduce conflicting updates, and cuts down on
+the need to respond to a conflict with a new update. Under this model, you won't
+have multiple users attempting to change data in primary product entry. At the
+very worst, you'll have multiple users attempting to claim a single ticket, and
+if you've grabbed several of those from your view, you simply move on to the
+next ticket and try again.
+
+.. note::
+    This FAQ entry was borrowed from
+    http://stackoverflow.com/questions/299723/can-i-do-transactions-and-locks-in-couchdb
+    with permission from the author
+
+Why does creating my view take so long?
```

Comment:
"Javascript view"

At some point we need to go through the documentation (and consult with user@ and dev@ mailing lists) to standardize on the jumble of JS View, Mango Index, and the more suitable generic term "secondary index".